sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,314 @@
+"""
+Implementation taken from Lemos et al, 'Sampling-Based Accuracy Testing of
+Posterior Estimators for General Inference' https://arxiv.org/abs/2302.03026
+
+The TARP diagnostic is a global diagnostic which can be used to check a
+trained posterior against a set of true values of theta.
+"""
+
+import warnings
+from typing import Callable, Optional, Tuple
+
+# import numpy as np
+import torch
+from joblib import Parallel, delayed
+from sbi.inference.posteriors.base_posterior import NeuralPosterior
+from sbi.inference.posteriors.vi_posterior import VIPosterior
+from sbi.simulators.simutils import tqdm_joblib
+from sbi.utils.metrics import l1, l2
+from scipy.stats import kstest
+from torch import Tensor
+from tqdm.auto import tqdm
+
+
+def infer_posterior_on_batch(
+    xs: Tensor,
+    posterior: NeuralPosterior,
+    num_posterior_samples: int = 1000,
+) -> Tensor:
+    """
+    Infer samples of a posterior distribution on a batch of inputs.
+
+    Parameters:
+    ----------
+    xs : Tensor
+        The input data batch.
+    posterior : NeuralPosterior
+        The neural posterior to use for inference.
+    num_posterior_samples : int, optional
+        The number of posterior samples to draw for each input, by default 1000.
+
+    Returns:
+    -------
+    Tensor
+        A tensor of shape (num_posterior_samples, N, P) where N is the number of
+        samples given by xs and P is the output dimension of the neural
+        posterior estimator.
+    """
+
+    samples = []
+
+    for idx in range(xs.shape[0]):
+        # unsqueeze for potential higher-dimensional data.
+        xo = xs[idx].unsqueeze(0)
+        # VI posterior needs to be trained on the current xo.
+        if isinstance(posterior, VIPosterior):
+            posterior.set_default_x(xo)
+            posterior.train()
+
+        # Draw posterior samples and save one for the data average posterior.
+        ths = posterior.sample((num_posterior_samples,), x=xo, show_progress_bars=False)
+        # Note: one could calculate coverage values here
+
+        samples.append(ths.unsqueeze(1))
+
+    return torch.cat(samples, dim=1)
+
+
+# this function currently does not perform any TARP related operation
+# the purpose of the function is (a) to align with the sbc interface and
+# (b) to provide the data which is required to run TARP
+def prepare_estimates(
```

Comment:
Please see comment below: I suggest to drop this function entirely and just have one loop in the main tarp function that samples from the posterior and directly calculates the desired distances etc.

Reply:
please just leave out the parallelization entirely for now, just assert `num_workers=1` and raise a `NotImplementedError`. 
We are working on refactoring the multiprocessing with `joblib` and `tqdm` in a different PR.

Reply:
As discussed, I'll keep it as is for the time being. I added a note, that his function will be removed in the future. 90bb7705